### ISSUE TITLE
Clang Static Analysis

### ISSUE THREAD

**Issue by hjmjohnson** [stnava/ANTs#275]
@ntustison @stnava 

Nick and Brian,

I ran ANTs through clang's static analysis tool.  It identified 7 bugs, and 20 warnings.    The attached files are reports of what was found.

If you replace my source path with your source path, you should be able to use 

vim -q ANTS_FULL_SCAN_ANALYSIS.txt to step through each reported error.

[ANTS_SHORT_SCAN_ANALYSIS.txt](https://github.com/stnava/ANTs/files/102505/ANTS_SHORT_SCAN_ANALYSIS.txt)
[ANTS_FULL_SCAN_ANALYSIS.txt](https://github.com/stnava/ANTs/files/102504/ANTS_FULL_SCAN_ANALYSIS.txt)

============================
If you let me know when some of these are resolved, I can create new reports.

Hans

**Comment by gdevenyi** [comment]
Huh, I guess I can close #258

**Comment by ntustison** [comment]
Thanks very much, Hans.  I'll start working on these errors tomorrow.

**Comment by ntustison** [comment]
@stnava, what do you think about getting rid of ExternalApplications/MALF?  That's causing a number of clang issues.

**Comment by stnava** [comment]
that's a great idea.


brian

**Comment by ntustison** [comment]
Okay, I'll work on that.  In addition to MALF, there's that learning stuff but I think all that we would do that type of thing through ANTsR anyway.

**Comment by ntustison** [comment]
Okay, @hjmjohnson , can you re-run the analysis when you get a chance?  Thanks.

**Comment by hjmjohnson** [comment]
@ntustison There are still some very concerning warnings:  In order of most conerning o least concerning:

/scratch/johnsonhj/src/BT-scanbuild/ANTs/ImageSegmentation/antsManifoldParzenWindowsListSampleFunction.hxx:49:1: warning: all paths through this function will call itself [-Winfinite-recursion]
/scratch/johnsonhj/src/BT-scanbuild/ANTs/ImageSegmentation/antsJointHistogramParzenShapeAndOrientationListSampleFunction.hxx:401:7: warning: variable 'tIter' is incremented both in the loop header and in the loop body [-Wfor-loop-analysis]
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ImageMath.cxx:1080:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/AverageImages.cxx:80:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/AverageImages.cxx:197:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:306:30: warning: Use of zero-allocated memory
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:297:28: warning: Use of zero-allocated memory
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:287:5: warning: Value stored to 'tokenPos' is never read
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:286:5: warning: Value stored to 'delimPos' is never read

**Comment by ntustison** [comment]
Thanks @hjmjohnson.  I didn't address them all during the first pass.  I should be able to finish the rest this week.

**Comment by hjmjohnson** [comment]
Some should be really easy.  

i.e.:
warning: variable 'tIter' is incremented both in the loop header and in the loop body
I think you just want to remove teh tIter++ at the bottom of the loop.

warning: Value stored warnings is to just delete the variable, or remove setting it when the value is not used (i.e. don't initialize to zero, and then on the next line set it to something else).

**Comment by ntustison** [comment]
Yeah, I just realized that.  I'll finish it up tonight.

**Comment by hjmjohnson** [comment]
The
warning: The right operand of '>' is a garbage value
 
can be hard to track down.  It is likely an uninitialized variable some where up the call stack under certain conditions.

void foo(int x) {
 if( x > 1) {} /* x could be garbage here */
}

int y;  /* <-- Not initialized */
if ( rand() >  0.6 ) { y = 1; }
else {  /* not setting y */ }

**Comment by ntustison** [comment]
Thanks Hans.  I hadn't ever seen that warning so I appreciate the help.  I think I made some progress with the most recent commit.

**Comment by ntustison** [comment]
I think this was addressed.  @hjmjohnson can reopen if the issue was not addressed.

**Comment by hjmjohnson** [comment]
@ntustison Nick, only a first pass was made on this.  There are still several of the warnings that are stil outstanding.
@ntustison There are still some very concerning warnings: In order of most conerning o least concerning:

/scratch/johnsonhj/src/BT-scanbuild/ANTs/ImageSegmentation/antsManifoldParzenWindowsListSampleFunction.hxx:49:1: warning: all paths through this function will call itself [-Winfinite-recursion]
/scratch/johnsonhj/src/BT-scanbuild/ANTs/ImageSegmentation/antsJointHistogramParzenShapeAndOrientationListSampleFunction.hxx:401:7: warning: variable 'tIter' is incremented both in the loop header and in the loop body [-Wfor-loop-analysis]
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ImageMath.cxx:1080:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/AverageImages.cxx:80:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/AverageImages.cxx:197:37: warning: The right operand of '>' is a garbage value
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:306:30: warning: Use of zero-allocated memory
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:297:28: warning: Use of zero-allocated memory
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:287:5: warning: Value stored to 'tokenPos' is never read
/scratch/johnsonhj/src/BT-scanbuild/ANTs/Examples/ANTS.cxx:286:5: warning: Value stored to 'delimPos' is never read

**Comment by ntustison** [comment]
Thanks for this, @hjmjohnson .  I'll take a look today and see if I can resolve a couple more.

**Comment by ntustison** [comment]
Hey @hjmjohnson , can you recheck those warnings you just sent?  I thought I addressed many of them in this [commit](https://github.com/stnava/ANTs/commit/46a1bb1faadbaef7f8e3ee0ff27cfc8b81069e5a#diff-1217471a3b83d7f26f9cf0edc39a8f2d).

**Comment by hjmjohnson** [comment]
I’m running a fresh copy now.

It won’t be done until tomorrow.


Hans

--

**Comment by ntustison** [comment]
Oh, no hurry.  Thanks for doing it.

**Comment by hjmjohnson** [comment]
@ntustison Only a few items left.  I missed the patch that addressed some of the previous errors.  Sorry.

[warning_logger.txt](https://github.com/stnava/ANTs/files/156782/warning_logger.txt)

**Comment by ntustison** [comment]
@hjmjohnson, I attempted a fix [here](https://github.com/stnava/ANTs/commit/72c5e9a2a2ec3ba7920d7705c6345437d771c2ec) but I'm not sure if those changes will clear up those three ANTs-specific warnings.  I searched for this type of warning but, even then, it wasn't clear to me what we're doing wrong in these three cases.    I'm not going to worry about the ITK-specific ones as I'm guessing somebody else has or will address them.
